keps/sig-storage/2605-option-to-preserve-default-file-mode-bit-in-atomicwriter-volumes/README.md

```diff
@@ -0,0 +1,181 @@
+
+With this feature, we try to provide a backwards compatible way for the users to
+to have a tighter/predictable control over the file mode set on the files in AtomicWriter volumes.
+
+We intend to provide a way for the users to opt-in for the new behavior and expect not to break
+any existing applications/configurations.
+
+## Proposal
+
+Kubernetes should implement a new boolean flag (PreservePermissions) in the SecretVolumeSource and
+other AtomicWriter volumes.
+
+The flag should be optional and default to false in code. This flag should be honored at the time of volume
+set-up and call to SetVolumeOwnership should be skipped in case PreservePermissions=true.
+
+### Preserve Permissions Flag
+
+A new PreservePermissions boolean will be implemented in SecretVolumeSource and other AtomicWriter volumes.
+
+```go
+// SecretVolumeSource adapts a Secret into a volume.
+//
+// The contents of the target Secret's Data field will be presented in a volume
+// as files using the keys in the Data field as the file names.
+// Secret volumes support ownership management and SELinux relabeling.
+type SecretVolumeSource struct {
+    // Name of the secret in the pod's namespace to use.
+    // +optional
+    SecretName string
+    // If unspecified, each key-value pair in the Data field of the referenced
+    // Secret will be projected into the volume as a file whose name is the
+    // key and content is the value. If specified, the listed keys will be
+    // projected into the specified paths, and unlisted keys will not be
+    // present. If a key is specified which is not present in the Secret,
+    // the volume setup will error unless it is marked optional. Paths must be
+    // relative and may not contain the '..' path or start with '..'.
+    // +optional
+    Items []KeyToPath
+    // Mode bits to use on created files by default. Must be a value between
+    // 0 and 0777.
+    // Directories within the path are not affected by this setting.
+    // This might be in conflict with other options that affect the file
+    // mode, like fsGroup, and the result can be other mode bits set.
+    // If PreservePermissions is set to true then file mode bit is not
+    // modified after the file creation, due to any other options set
+    // such as fsGroup etc
+    // +optional
+    DefaultMode *int32
+    // Specify whether the Secret or its key must be defined
+    // +optional
+    Optional *bool
+    // Specify whether we want to preserve the DefaultMode set on the files
+    // at the time of file creation.
+    // If PreservePermissions is set to true the file mode is not modified based
+    // on fsGroup after the file is created with the defaultMode.
+    // If the PreservePermissions is set to false and fsGroup is specified
+    // the volume is modified to be owned by the fsGroup.
+    // If PreservePermissions is not specified in the volume spec, it defaults to false.
+    // +optional
+    PreservePermissions *bool
```

Comment:
Now that I am looking at the API, would it make sense to move the flag into [`struct Volume`](https://github.com/kubernetes/kubernetes/blob/01289178ea58df0520ec9cea0ce4bdc8aa6e93f1/pkg/apis/core/types.go#L45)? This way, it could be used with any volume, incl. in-line ephemeral CSI volumes and, if someone really needs, to opt-out *any* volume in a pod from fsGroup.

I am not sure if it makes things easier or harder on kubelet side.